src/overlays/actors/ovl_En_Ex_Ruppy/z_en_ex_ruppy.c

```diff
@@ -21,29 +34,345 @@ const ActorInit En_Ex_Ruppy_InitVars = {
+        this->actor.speedXZ = 0.0f;
+        this->actor.velocity.z = 0.0f;
+        this->actor.velocity.y = 0.0f;
+        this->actor.velocity.x = 0.0f;
+        this->actor.gravity = 0.0f;
+        func_80078914(&this->actor.projectedPos, NA_SE_EV_BOMB_DROP_WATER);
+        this->actionFunc = EnExRuppy_ChoseFinalPos;
+    } 
+}
+
+void EnExRuppy_ChoseFinalPos(EnExRuppy *this, GlobalContext *globalCtx) {
+    s32 pad;
+    f32 temp_f2;
+    if (((DIVING_GAME != NULL) && (DIVING_GAME->actor.update != NULL)) && ((DIVING_GAME->unk_2A2 == 2))) {
+         this->isFalling = 0;
+         this->actor.posRot.pos.x = ((Math_Rand_ZeroOne() - 0.5f) * 300.0f) + -260.0f;
+         this->actor.posRot.pos.y = ((Math_Rand_ZeroOne() - 0.5f) * 200.0f) + 370.0f;
+         temp_f2 =  this->unk_15A * -50.0f;
+         if (!(gSaveContext.eventChkInf[3] & 0x100)) {
+             temp_f2 += -500.0f;
+             this->actor.posRot.pos.z = ((Math_Rand_ZeroOne() - 0.5f) * 80.0f) + temp_f2;
+         } else {
+             temp_f2 += -300.0f;
+             this->actor.posRot.pos.z = ((Math_Rand_ZeroOne() - 0.5f) * 60.0f) + temp_f2;
+         }
+         this->actionFunc = EnExRuppy_Sink;
+         this->actor.gravity = -1.0f;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ex_Ruppy/EnExRuppy_Destroy.s")
+void EnExRuppy_Sink(EnExRuppy *this, GlobalContext *globalCtx) {
+    s32 pad;
+    Vec3f tempPos;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ex_Ruppy/func_80A0A820.s")
+    if ((this->actor.bgCheckFlags & 0x20) && (15.0f < this->actor.waterY)) {
+        tempPos = this->actor.posRot.pos;
+        tempPos.y += this->actor.waterY;
+        this->actor.velocity.y = -1.0f;
+        this->actor.gravity = -0.2f;
+        func_8002949C(globalCtx, &tempPos, 0, 0, 0, 800);
+        func_80078914(&this->actor.projectedPos, NA_SE_EV_BOMB_DROP_WATER);
+        this->actionFunc = func_80A0AD88;
+    }
+    if ((DIVING_GAME != 0) && (DIVING_GAME->actor.update != 0) && (DIVING_GAME->unk_29C == 0)) {
+        this->timer = 20;
+        this->actionFunc = func_80A0AEE0;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ex_Ruppy/func_80A0AA3C.s")
+void func_80A0AD88(EnExRuppy *this, GlobalContext *globalCtx) {
+    EnDivingGame* divingGame;
+    Vec3f unusedVec3f;
+    Vec3f unusedVec3f2;
+    const f32 localConst = 30.0f;
+    unusedVec3f = D_80A0B388;
+    unusedVec3f2 = D_80A0B394;
+    if (this->timer == 0) {
+        this->timer = 10;
+        func_800293E4(globalCtx, &this->actor.posRot.pos, 0.0f, 5.0f, 5.0f, Math_Rand_ZeroFloat(0.03f) + 0.07f);
+    }
+    if (DIVING_GAME != NULL) {
+        divingGame = DIVING_GAME;
+        if (divingGame->actor.update != NULL) {
+            if (divingGame->unk_29C == 0) {
+                this->timer = 20;
+                this->actionFunc = func_80A0AEE0;
+                return;
+            }
+            if (this->actor.xyzDistFromLinkSq < SQ(localConst)) {
```

Comment:
probably else here

Reply:
I just tried it and it caused regalloc and ordering issues